multicore_runtime/middleware.py

```diff
@@ -0,0 +1,185 @@
+    # Wipe os.environ entirely.
+    os.environ.clear()
+
+    # Populate os.environ in order, so that later steps overwrite conflicting
+    # keys in previous steps.
+
+    # Add in user env variables specified in app.yaml. These are added first
+    # because they should not take precedence over any conflicting key.
+    os.environ.update(frozen_user_env)
+
+    # Repopulate os.environ with the frozen environment.
+    os.environ.update(frozen_environment)
+
+    # Add in wsgi_env data, including request headers.
+    os.environ.update(request_environment_for_wsgi_env(wsgi_env))
+
+    # Add in configuration data from env_config.
+    os.environ.update(frozen_env_config_env)
+
+    # Add reserved keys, which draw from wsgi_env as well. These have a very
+    # high priority and so are added nearly last.
+    os.environ.update(reserved_env_keys_for_wsgi_env(wsgi_env))
+
+    # Tweak the environment to hide the service bridge.
+    os.environ.update(get_env_to_hide_service_bridge(wsgi_env))
+
+    return app(wsgi_env, start_response)
+
+  return reset_environment_wrapper
+
+
+def request_environment_for_wsgi_env(wsgi_env):
+  """Get a dict of key-value pairs from wsgi_env that work as env variables.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  # Return all key, value pairs from wsgi_env where value is a string.
+  return {key: value for key, value in wsgi_env.iteritems()
+          if isinstance(value, basestring)}
+
+
+def reserved_env_keys_for_wsgi_env(wsgi_env):
+  """Get a dict for reserved keys based on wsgi_env headers and defaults.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  output = {}
+
+  # Use the default value for a reserved key if the corresponding header is not
+  # set, or if the header exists but its value is blank.
+  for key, default in RESERVED_ENV_KEYS.iteritems():
+    value = wsgi_env.get('HTTP_X_APPENGINE_' + key)
+    output[key] = value or default  # Must be set to a valid value or default.
+
+  return output
+
+
+def get_env_to_hide_service_bridge(wsgi_env):
+  """Generate a dictionary of environment variables to hide the service bridge.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+  output = {}
+
+  # Because the request is coming over the service bridge, the service bridge
+  # host and port that are automatically populated in SERVER_NAME and
+  # SERVER_PORT, respectively.  However, this is an implementation detail that
+  # should not be shown to user code. Instead, we'll rely on the HTTP Host
+  # header to retrieve the hostname used in the original request.  This mimics
+  # the behavior of non-VM App Engine.
+  if 'HTTP_HOST' in wsgi_env:
```

Comment:
Alternatively, use `wsgi_env.get('HTTP_HOST', None)` and check for `None` (or implicit false, if applicable, particularly if you don't care about empty string), that way you don't have two separate lookups, one for the `in` and the other for the actual access to the value...